File: 0844-backspace-string-compare/0844-backspace-string-compare.java
class Solution {
    public boolean backspaceCompare(String s, String t) {
      int idx1=s.length()-1,idx2=t.length()-1;
        while (idx1>=0||idx2>=0){
            int i1=getNextValidCharIdx(s,idx1);
            int i2=getNextValidCharIdx(t,idx2);
            if(i1<0&&i2<0){
                return true;
            }
            if(i1<0||i2<0){
                return false;
            }
            if(s.charAt(i1)!=t.charAt(i2)){
                return false;
            }
            idx1=i1-1;
            idx2=i2-1;
        }
        return true;

    }

    private int getNextValidCharIdx(String str, int index) {
        int backspaceCount=0;
        while (index>=0){
            if(str.charAt(index)=='#'){
                backspaceCount++;
            }else if(backspaceCount>0){
                backspaceCount--;
            }else{
                break;
            }
            index--;
        }
        return index;
    }
}

File: 0844-backspace-string-compare/0844-backspace-string-compare.py
class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
        i, j = len(s)-1, len(t)-1
        ss = tt = 0
        while 0 <= i or 0 <= j: 
            while 0 <= i and (s[i] == "#" or ss): 
                if s[i] == "#": ss += 1
                else: ss -= 1
                i -= 1
            while 0 <= j and (t[j] == "#" or tt): 
                if t[j] == "#": tt += 1
                else: tt -= 1
                j -= 1
            if i < 0 and 0 <= j or 0 <= i and j < 0 or 0 <= i and 0 <= j and s[i] != t[j]: return False 
            i, j = i-1, j-1
        return True 